omniscidb/QueryBuilder/QueryBuilder.h

```diff
@@ -0,0 +1,656 @@
+  BuilderExpr gt(int64_t val) const;
+  BuilderExpr gt(float val) const;
+  BuilderExpr gt(double val) const;
+  BuilderExpr gt(const std::string& val) const;
+
+  BuilderExpr ge(const BuilderExpr& rhs) const;
+  BuilderExpr ge(int val) const;
+  BuilderExpr ge(int64_t val) const;
+  BuilderExpr ge(float val) const;
+  BuilderExpr ge(double val) const;
+  BuilderExpr ge(const std::string& val) const;
+
+  BuilderExpr at(const BuilderExpr& idx) const;
+  BuilderExpr at(int idx) const;
+  BuilderExpr at(int64_t idx) const;
+
+  BuilderExpr rewrite(ExprRewriter& rewriter) const;
+
+  BuilderExpr operator!() const;
+  BuilderExpr operator-() const;
+  BuilderExpr operator[](const BuilderExpr& idx) const;
+  BuilderExpr operator[](int idx) const;
+  BuilderExpr operator[](int64_t idx) const;
+
+  const QueryBuilder& builder() const;
+  Context& ctx() const;
+
+ protected:
+  friend class QueryBuilder;
+  friend class BuilderNode;
+
+  BuilderExpr(const QueryBuilder* builder,
+              ExprPtr expr,
+              const std::string& name = "",
+              bool auto_name = true);
+
+  const QueryBuilder* builder_;
+  ExprPtr expr_;
+  std::string name_;
+  bool auto_name_;
+};
+
+class BuilderSortField {
+ public:
+  BuilderSortField(int col_idx,
+                   SortDirection dir,
+                   NullSortedPosition null_pos = NullSortedPosition::Last);
+  BuilderSortField(int col_idx,
+                   const std::string& dir,
+                   const std::string& null_pos = "last");
+  BuilderSortField(const std::string& col_name,
+                   SortDirection dir,
+                   NullSortedPosition null_pos = NullSortedPosition::Last);
+  BuilderSortField(const std::string& col_name,
+                   const std::string& dir,
+                   const std::string& null_pos = "last");
+  BuilderSortField(BuilderExpr expr,
+                   SortDirection dir,
+                   NullSortedPosition null_pos = NullSortedPosition::Last);
+  BuilderSortField(BuilderExpr expr,
+                   const std::string& dir,
+                   const std::string& null_pos = "last");
+
+  bool hasColIdx() const { return std::holds_alternative<int>(field_); }
+  bool hasColName() const { return std::holds_alternative<std::string>(field_); }
+  bool hasExpr() const { return std::holds_alternative<BuilderExpr>(field_); }
+
+  int colIdx() const { return std::get<int>(field_); }
+  std::string colName() const { return std::get<std::string>(field_); }
+  ExprPtr expr() const { return std::get<BuilderExpr>(field_).expr(); }
+  SortDirection dir() const { return dir_; }
+  NullSortedPosition nullsPosition() const { return null_pos_; }
+
+  static SortDirection parseSortDirection(const std::string& val);
+  static NullSortedPosition parseNullPosition(const std::string& val);
+
+ protected:
+  std::variant<int, std::string, BuilderExpr> field_;
+  SortDirection dir_;
+  NullSortedPosition null_pos_;
+};
+
+class BuilderNode {
```

Comment:
Could you please elaborate on the abstractions' relation to each other and the reasoning behind them?

Reply:
`BuilderNode` wraps `hdk::ir::Node` and provides an additional building interface to it. `BuilderExpr` makes the same for `hdk::ir::Expr`.